session_test.go

```diff
@@ -67,14 +68,152 @@ func (p *pipeConn) Close() error {
 	return p.writer.Close()
 }
 
-func testConn() (io.ReadWriteCloser, io.ReadWriteCloser) {
+func testConnPipe(testing.TB) (io.ReadWriteCloser, io.ReadWriteCloser) {
 	read1, write1 := io.Pipe()
 	read2, write2 := io.Pipe()
 	conn1 := &pipeConn{reader: read1, writer: write2}
 	conn2 := &pipeConn{reader: read2, writer: write1}
 	return conn1, conn2
 }
 
+func testConnTCP(t testing.TB) (io.ReadWriteCloser, io.ReadWriteCloser) {
+	l, err := net.ListenTCP("tcp", nil)
+	if err != nil {
+		t.Fatalf("error creating listener: %v", err)
+	}
+	t.Cleanup(func() { _ = l.Close() })
+
+	network := l.Addr().Network()
+	addr := l.Addr().String()
+
+	var server net.Conn
```

Comment:
This feels slightly racy. Probably better to send this `net.Conn` back out of the goroutine using another channel?

Reply:
`server` is only used after receiving on `errCh` so errCh synchronizes it. (CI also runs with `-race`, and I did things like `go test -race -count=1000` locally when trying desperately to find a root cause of what was seen in #127.)

Reply:
Oh man I looked right over `defer close(errCh)` while skimming for `errCh <- nil` :facepalm: